woocommerce/woocommerce-ios
# [Mobile Payments] [SDK 2] Allow collect payment on orders where payment collection was interrupted by an error

Closes #4655

Changes:
- If collecting a payment fails and the payment method ID is changed to `woocommerce_payments` the logic prior to this PR would treat it as ineligible for in-person payments
- This PR adds `woocommerce_payments` as an acceptable payment method ID on an order. Coupled with the status of processing, on-hold, or none makes sense and continues to avoids the problem of collecting payments on orders which already have payment collected
- This PR also presents any error message provided from the backend during order payment capture directly to the user (e.g. "Payment capture failed to complete with the following message: Error: Invalid integer: 2.5") if preferred we could just show something simpler like "Payment capture failed to complete" - it would also avoid the risk of localization differences but would impart less information to the user. 

To test:
- Use the fractional quantity plugin described in @designsimply 's original report in #4655 to reproduce the error during payment collection
- Ensure that after pulling to refresh the order the Collect Payment button continues to be presented

Note: The backend does NOT emit different codes for different capture failures from Stripe, however. They all have the same code: `wcpay_capture_error`. See https://github.com/Automattic/woocommerce-payments/blob/6b2e9a5c9f372fdf9d585220846ac0154e83dea6/includes/admin/class-wc-rest-payments-orders-controller.php#L125-L139

Note: We could add coverage for other errors like wcpay_payment_uncapturable and wcpay_refunded_order_uncapturable and the less likely wcpay_missing_order  if desired. I can open separate issues for any of those. See that same part of the plugin linked above for those. See below for a screenshot of how other errors are presented (i.e. no change)

Screenshots:

Error when we attempt to process a fractional quantity:

![IMG_0204](https://user-images.githubusercontent.com/1595739/137977158-a0c8d023-a84f-458a-b48c-fd97539e1b30.PNG)

Error when the endpoint simply doesn't exist (edit includes/admin/class-wc-rest-payments-orders-controller.php and rename the path):

![IMG_0205](https://user-images.githubusercontent.com/1595739/137977148-b656fa14-a221-49ce-a959-a05dad359d2a.PNG)

Update release notes:
